category/execution/ethereum/trace/prestate_tracer.cpp

```diff
@@ -107,21 +109,87 @@ namespace trace
         state_deltas_to_json(state_deltas, state, storage_);
     }
 
-    void run_tracer(StateTracer const &tracer, State &state)
+    AccessListTracer::AccessListTracer(
+        nlohmann::json &storage, Address const &sender,
+        Address const &beneficiary, std::optional<Address> const &to,
+        std::span<std::optional<Address> const> const authorities)
+        : storage_(storage)
     {
-        if (std::holds_alternative<PrestateTracer>(tracer)) {
-            PrestateTracer prestate = std::get<PrestateTracer>(tracer);
-            prestate.encode(state.original(), state);
-            return;
+        excluded_addresses_.insert(sender);
+        excluded_addresses_.insert(beneficiary);
+
+        if (to.has_value()) {
+            excluded_addresses_.insert(*to);
+        }
+
+        for (auto const &authority : authorities) {
+            if (authority.has_value()) {
+                excluded_addresses_.insert(*authority);
+            }
         }
+    }
+
+    template <Traits traits>
+    void AccessListTracer::encode(State &state)
+    {
+        auto access_list = json::array();
+        for (auto const &[address, current_stack] : state.current()) {
+            auto keys = json::array();
+            auto const &current_account_state = current_stack.recent();
+            for (auto const &key :
+                 current_account_state.get_accessed_storage()) {
+                keys.push_back(bytes_to_hex(key.bytes));
+            }
+
+            // If an address is excluded because it's always considered warm, we
+            // still want to include it in the access list if it's had storage
+            // keys set by this transaction.
+            auto const exclude =
+                keys.empty() && should_exclude_address<traits>(address);
+
+            if (!exclude) {
+                auto entry = json::object();
+                entry["address"] = bytes_to_hex(address.bytes);
+                entry["storageKeys"] = std::move(keys);
```

Comment:
Alternatively, we can have the object initializer populate the entry, thereby eliminating the need for mutation (probably needs to be run through the clang formatter first).
```suggestion
                auto const entry = json::object({
                     {"address", bytes_to_hex(address.bytes)},
                     {"storageKeys", std::move(keys)}
                 });
```